### PR TITLE
Package the HIP/ROCm and ROCrand libraries

### PR DESCRIPTION
**Pull request by fwyzard**
Package the core HIP/ROCm libraries and tools, and the ROCrand libraries.
Update the tool files for rocm-rocrand and alpaca-rom accoridngly.

### CODE REVIEW

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
```
Trying to install the rpm package external+rocm-rocrand+5.4.3-d6c59ff4834c2aa5f7e34b44613e08e8 just built.
Checking local path dependency for rpm package external+rocm-rocrand+5.4.3-d6c59ff4834c2aa5f7e34b44613e08e8 just build.
Failed to install RPM for rocm-rocrand
```

Is there a way to get more details about why the installation failed?

File: rocm-rocrand.spec [hunk 1]
```diff
@@ -0,0 +1,17 @@
+### RPM external rocm-rocrand 5.4.3
+## NOCOMPILER
+Source0: https://repo.radeon.com/rocm/rhel%{rhel}/%{realversion}/main/rocrand-2.10.9.50403-121.el%{rhel}.x86_64.rpm
+Source1: https://repo.radeon.com/rocm/rhel%{rhel}/%{realversion}/main/rocrand-devel-2.10.9.50403-121.el%{rhel}.%{_arch}.rpm
+
```
Comment (fwyzard) [inline rocm-rocrand.spec:16]: maybe this should have
```
Requires: rocm
```

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@smuzaffar, I have a couple of questions:
  - `rocgdb` (the ROCm version of `gdb`) depends on `libbabeltrace-1`, which we do not package; is it OK to add this extra dependency to cmsdist/cmssw ?
  - `rocgdb` depends on `libdebuginfod.so.1` and `libgmp.so.10`, which are part of our gcc packages; how do I declare a dependency on it ?
  - something depends on `perl(File::Which)` and `perl(URI::Encode)`; do we add them to the fakesystem spec ?

**Comment by fwyzard** [comment]
- the AMD HSA runtime depends on the AMD drivers... but I have absolutely no experience with them, and I do not know if adding them to our packages is a good idea or not :-(

**Comment by fwyzard** [comment]
> * the AMD HSA runtime depends on the AMD drivers... but I have absolutely no experience with them, and I do not know if adding them to our packages is a good idea or not :-(

Correction: it depends on the `/usr/lib64/libdrm*` libraries, but those should be preinstalled on all (most ?) systems. In fact they are a dependency of `mesa-libGL`, which is a dependency we already have.

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test

**Comment by smuzaffar** [comment]
>rocgdb (the ROCm version of gdb) depends on libbabeltrace-1, which we do not package; is it OK to add this extra dependency 
to cmsdist/cmssw ?

sure, this is OK.

> rocgdb depends on libdebuginfod.so.1 and libgmp.so.10, which are part of our gcc packages; how do I declare a dependency on it ?

I think you can remove `## NOCOMPILER` and then `cmsBuild` will automatically add `gcc` dependency

>something depends on perl(File::Which) and perl(URI::Encode); do we add them to the fakesystem spec ?

yes this is fine. Make sure to update the revision of thepackage i.e. increment `## REVISION 1012` in fakesystem.spec

File: fakesystem.spec [hunk 2]
```diff
@@ -107,6 +107,10 @@ Provides: perl(sort)
 # ROCM
 #####
 Provides: libamdhip64.so.5(hip_5.3)(64bit)
+Provides: libdrm.so.2()(64bit)
+Provides: libdrm_amdgpu.so.1()(64bit)
+Provides: perl(File::Which)
+Provides: perl(URI::Encode)
```
Comment (smuzaffar) [inline fakesystem.spec:113]: @fwyzard , for now I would suggest to add these to the package which needs these. Once we have a working PR then we can move these requires in to fakesystem and update the revision

Comment (fwyzard) [reply fakesystem.spec:113]: OK, I will do it like that for the local tests, and then push them here once they work locally.

Comment (fwyzard) [reply fakesystem.spec:113]: How can I test the `fakesystem` changes locally ?

Comment (smuzaffar) [reply fakesystem.spec:113]: just build `fakesystem` first

Comment (fwyzard) [reply fakesystem.spec:113]: ah, ok !

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
I've dropped roc-gdb for the time being, because it seems to depend on the _system_ version of the python libraries (3.6m).
I would anyway expect that any actual ROCm installation provides it from the system, so we shouldn't need to bundle it ourselves...

**Comment by fwyzard** [comment]
> -1
> 
> **Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7f7dbd/31190/summary.html **COMMIT**: [efbe8fc](https://github.com/cms-sw/cmsdist/commit/efbe8fc194f119811839d6e0bb35f94448116ee0) **CMSSW**: CMSSW_13_1_X_2023-03-08-2300/el9_amd64_gcc11 **User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmsdist/8370/31190/install.sh` to create a dev area with all the needed externals and cmssw changes.
> ## External Build
> 
> I found compilation warning when building: See details on the summary page.

The error I see in the log is
```
Error while downloading https://repo.radeon.com/rocm/rhel%{rhel}/5.4.3/main/comgr-2.4.0.50403-121.el%{rhel}.x86_64.rpm: HTTP Error 400: Bad Request
Error while downloading https://repo.radeon.com/rocm/rhel%{rhel}/5.4.3/main/hip-devel-5.4.22804.50403-121.el%{rhel}.x86_64.rpm: HTTP Error 400: Bad Request
...
```

But, for example, I see that https://repo.radeon.com/rocm/rhel9/5.4.3/main/comgr-2.4.0.50403-121.el9.x86_64.rpm exists.

@smuzaffar. do you know if `%{rhel}` does not get properly expanded to `9` on el9 ?

**Comment by fwyzard** [comment]
> @smuzaffar. do you know if `%{rhel}` does not get properly expanded to `9` on el9 ?

I seems work by itself:
```
$ cmssw-el9
...
$ rpm --eval '%{rhel}'
9
```

**Comment by smuzaffar** [comment]
please test for el9_amd64_gcc11

@fwyzard , `cmsBuild` was using only `/etc/rpm/macros.dist` for distro macros. on `el9` it is available via `/usr/lib/rpm/macros.d/macros.dist`. I have updated `cmsBuild` to make use of both file. I have restart the tests which should pick up the latest pkgtools

**Comment by fwyzard** [comment]
Thanks !

**Comment by fwyzard** [comment]
Looks like [that](https://github.com/cms-sw/cmsdist/pull/8370#issuecomment-1463789348) actually broke the el9 build ?

**Comment by smuzaffar** [comment]
> Looks like [that](https://github.com/cms-sw/cmsdist/pull/8370#issuecomment-1463789348) actually broke the el9 build ?

humm strange
- system `rpm2cpio` works for `comgr-2.4.0.50403-121.el9.x86_64.rpm`
- cms `rpm2cpio` (from `el9_amd64_gcc11/external/rpm/4.17.0-e83decb517178867c753ca12a66bdf0e/bin/rpm2cpio`) fails for  `comgr-2.4.0.50403-121.el9.x86_64.rpm`
- cms `rpm2cpio` works for RPM  build by cms

**Comment by smuzaffar** [comment]
@fwyzard , I found https://bugzilla.redhat.com/show_bug.cgi?id=2058426 . So looks like for el9 we need to do
```
rpm2cpio %{SOURCE0} | zstd -d | cpio -idmv
```

**Comment by fwyzard** [comment]
> @fwyzard , I found https://bugzilla.redhat.com/show_bug.cgi?id=2058426 . So looks like for el9 we need to do
> 
> ```
> rpm2cpio %{SOURCE0} | zstd -d | cpio -idmv
> ```

Mhm... can we update our version of RPM, at least for el9 ?

**Comment by smuzaffar** [comment]
we are using newer version of RPM for el9  (`4.17.0` as compare to `4.15.0` for `el8`)

**Comment by fwyzard** [comment]
> we are using newer version of RPM for el9 (`4.17.0` as compare to `4.15.0` for `el8`)

Can we enable the zstd support ?

From a quick look, it should be enough to add
```
BuildRequires: zstd
```
(or `Requires: zstd`? not sure about the difference here) and pass `--enable-zstd=yes`.

**Comment by smuzaffar** [comment]
> Can we enable the zstd support ?

sure we can, let me rebuilt rpm. We need to make sure that rpm with zstd support can still unpack/install RPMs we have built without zstd.

**Comment by smuzaffar** [comment]
please test for el9_amd64_gcc11

**Comment by smuzaffar** [comment]
@fwyzard , rpm for el9 has been updated to version 4.18.0 with zstd support.

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
please test for el8_amd64_gcc11

**Comment by fwyzard** [comment]
> -1
> 
> **Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7f7dbd/31304/summary.html **COMMIT**: [bdbc7ba](https://github.com/cms-sw/cmsdist/commit/bdbc7ba99c4b49c93d731bf710483ef52fb496ec) **CMSSW**: CMSSW_13_1_X_2023-03-15-2300/el9_amd64_gcc11 **User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmsdist/8370/31304/install.sh` to create a dev area with all the needed externals and cmssw changes.
> ## External Build
> 
> I found compilation error when building:
> ```
> File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/./pkgtools/cmsBuild", line 3651, in installPackage
> installRpm(pkg, pkg.options.bootstrap, scheduler)
> File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/./pkgtools/cmsBuild", line 3399, in installRpm
> raise RpmInstallFailed(pkg, output)
> RpmInstallFailed: Failed to install package rocm. Reason:
> error: Failed dependencies:
> 	perl(FindBin) is needed by external+rocm+5.4.3-1e47a6adebbf52caf2553c8ef70a80dc-1-1.x86_64
> 	perl(Hash::Util) is needed by external+rocm+5.4.3-1e47a6adebbf52caf2553c8ef70a80dc-1-1.x86_64
> 	perl(Sys::Hostname) is needed by external+rocm+5.4.3-1e47a6adebbf52caf2553c8ef70a80dc-1-1.x86_64
> 
> * The action "build-cms+cmssw-tool-conf+55.0-ca517681a0cd882a74fe09aa9a684c1e" was not completed successfully because The following dependencies could not complete:
> ```

@smuzaffar I thought I added these three dependencies to `fakesystem.spec`.
Any idea why it didn't work ?

**Comment by smuzaffar** [comment]
@fwyzard , remember that whenever you add stuff in fakesystem then you need to update the revision of the package too.

**Comment by fwyzard** [comment]
I had changed it as part of this PR.
Do I have to change it again if I update the PR ?

**Comment by smuzaffar** [comment]
as your changes has been built and uploaded to th PR repo so `cmsBuild` is always going to download the previous revision. So yes you need to update it again

**Comment by fwyzard** [comment]
ah... understood

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
OK, the actual error is
```
/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include/alpaka/rand/RandUniformCudaHipRand.hpp:34:22: fatal error: 'hiprand/hiprand_kernel.h' file not found
#            include <hiprand/hiprand_kernel.h>
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated when compiling for gfx900.
```

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@smuzaffar @makortel @pcanal looks like `genreflex` does not parse properly the `-isystem` directories.
Any ideas ?

**Comment by smuzaffar** [comment]
@fwyzard , I do not think it is `-isystem` issue , we have many externals with `-isystem` [a]. Also `DataFormats/PortableTestObjects/src/alpaka/classes_rocm.h` in IBs works fine ... right?

[a]
```
-isystem<path>/boost/1.80.0-415f9cccb73571663e34180e5839ce86/include
-isystem<path>/root/6.26.11-352ff7105472dc3a16fb5cfd35093c4a/include
-isystem<path>/tbb/v2021.8.0-d98ea1c603fa3368d06780eab09b1de9/include
```

**Comment by smuzaffar** [comment]
looks like scram did not add `rocm-rocrand` dependency in this case. I am looking in to it

**Comment by fwyzard** [comment]
Mhm, this is the command line that actually fails for me:
```
/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/lcg/root/6.26.11-149c0390ac099fe3c8283ca35bade3b4/bin/genreflex src/DataFormats/PortableTestObjects/src/alpaka/classes_rocm.h -s src/DataFormats/PortableTestObjects/src/alpaka/classes_rocm_def.xml -o tmp/el8_amd64_gcc11/src/DataFormats/PortableTestObjects/src/alpaka/DataFormatsPortableTestObjectsROCmAsync/a/DataFormatsPortableTestObjectsROCmAsync_xr.cc --fail_on_warnings --rootmap=tmp/el8_amd64_gcc11/src/DataFormats/PortableTestObjects/src/alpaka/DataFormatsPortableTestObjectsROCmAsync/a/DataFormatsPortableTestObjectsROCmAsync_xr.rootmap --rootmap-lib=libDataFormatsPortableTestObjectsROCmAsync.so -m DataFormatsCommon_xr_rdict.pcm -m DataFormatsProvenance_xr_rdict.pcm -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DCMS_DICT_IMPL -D_REENTRANT -DGNUSOURCE -D__STRICT_ANSI__ -DALPAKA_ACC_GPU_HIP_ENABLED -DALPAKA_HOST_ONLY -DGNU_GCC -D_GNU_SOURCE -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DEIGEN_DONT_PARALLELIZE -DTBB_USE_GLIBCXX_VERSION=110201 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH="CMSSW_13_1_X_2023-03-16-2300" -DPROJECT_NAME="CMSSW" -DPROJECT_VERSION="CMSSW_13_1_X_2023-03-16-2300" -I/data/user/fwyzard/CMSSW_13_1_X_2023-03-16-2300/src -I/data/user/fwyzard/CMSSW_13_1_X_2023-03-16-2300/poison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc11/cms/cmssw/CMSSW_13_1_X_2023-03-16-2300/src -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/pcre/8.43-5dcc901acc02f624b22dd9840b2357e8/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/boost/1.80.0-165bd6bfbfb4303b695a8c5946a5b92e/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/bz2lib/1.0.6-2c1f18484cb66c30aba7929f2be5e7d4/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/libuuid/2.34-0451b31e1b9a58c6aeefab41c18eea34/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm-rocrand/5.4.3-5cda51a0a8ef64a6f63f35d330d3072d/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/lcg/root/6.26.11-149c0390ac099fe3c8283ca35bade3b4/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/tbb/v2021.8.0-75c29f2d261a174db9e74a1b202d0dd2/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/xz/5.2.5-83d0a00b575efd1701e07bedf7977343/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/zlib/1.2.11-3dfb2715f3608466b74431b80eb9d788/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/eigen/82dd3710dac619448f50331c1d6a35da673f764a-9ac4aed18ac60d0189693c592862694d/include/eigen3 -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/fmt/8.0.1-43b841663c2a0d6622910a1ad66d228d/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/md5/1.0.0-e68283f2de2e2e709a0db99db3b53205/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm/5.4.3-e875674b47f88526f6f9c4594f5cafdd/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/tinyxml2/6.2.0-c2bad61e58f94d6db8f640afbd739be2/include -DCMSSW_REFLEX_DICT
```

The error I get is
```
In file included from input_line_7:77:
In file included from /data/user/fwyzard/CMSSW_13_1_X_2023-03-16-2300/src/DataFormats/Portable/interface/Product.h:7:
In file included from /cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include/alpaka/alpaka.hpp:29:
In file included from /cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include/alpaka/acc/AccGpuHipRt.hpp:14:
In file included from /cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include/alpaka/acc/AccGpuUniformCudaHipRt.hpp:26:
/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include/alpaka/rand/RandUniformCudaHipRand.hpp:34:22: fatal error: 'hiprand/hiprand_kernel.h' file not found
#            include <hiprand/hiprand_kernel.h>
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
Error: rootcling: compilation failure (tmp/el8_amd64_gcc11/src/DataFormats/PortableTestObjects/src/alpaka/DataFormatsPortableTestObjectsROCmAsync/a/DataFormatsPortableTestObjectsROCmAsync_xr4a71645ce4_dictUmbrella.h)
```

If I change `-isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm-rocrand/5.4.3-5cda51a0a8ef64a6f63f35d330d3072d/include` to `-I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm-rocrand/5.4.3-5cda51a0a8ef64a6f63f35d330d3072d/include`, then it works (though I get some warnings from rocrand headers, which is why I used `-isystem` in the first place).

That's why I though `-isystem` headers are treated differently.

**Comment by smuzaffar** [comment]
ah right, actually I do not see `-isystem<path>/rocm-rocrand/..` in the command generated by scram. So I need to debug why scram did not add this. But yes you are right , if I hadd `-isystem<path>/rocm-rocrand/..` by hand then `genreflex` fails but works with `-I<path>/rocm-rocrand/..`

**Comment by fwyzard** [comment]
If I modify the `rocm-rocrand.xml` tool file:
```diff
 <tool name="rocm-rocrand" version="5.4.3-5cda51a0a8ef64a6f63f35d330d3072d">
   <info url="https://github.com/ROCmSoftwarePlatform/rocRAND"/>
   <use name="rocm"/>
   <lib name="hiprand"/>
   <lib name="rocrand"/>
   <client>
     <environment name="ROCM_ROCRAND_BASE" default="/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm-rocrand/5.4.3-5cda51a0a8ef64a6f63f35d330d3072d"/>
     <environment name="LIBDIR"            default="$ROCM_ROCRAND_BASE/lib"/>
     <environment name="INCLUDE"           default="$ROCM_ROCRAND_BASE/include"/>
   </client>
-  <!-- use -isystem instead of -I to silence warnings in the HIP/ROCm headers -->
-  <flags SYSTEM_INCLUDE="1"/>
+  <flags CXXFLAGS="-Wno-unused-result"/>
 </tool>
```

and rebuild everything, so far it seems to work.

**Comment by smuzaffar** [comment]
@vgvassilev , do you know if `genreflex` works with `-isystem` ?

**Comment by fwyzard** [comment]
(and by the way, I have no idea why this breaks with respect to before where we had `rocm` and `rocm-rocrand` in a single package using `-isystem` 🤷🏻‍♂️ )

**Comment by smuzaffar** [comment]
I am also trying to understand that :-)

**Comment by smuzaffar** [comment]
@vgvassilev , the genreflex command [a] shows that it is not passing all the include directories to `rootcling` ( see https://muzaffar.web.cern.ch/genreflex-debug.txt for detail output).   looks like `genreflex` is not passing `isystemDir` to `rootcling`.

@fwyzard , In IB it works as rocm/include (where both rocm and rocm-rocrand headers are in same path) is available via ROOT_INCLUDE_PATH .  One way to fix this is to also add the following in rocm-rocrand.xml
```
<runtime name="ROOT_INCLUDE_PATH" value="$INCLUDE" type="path"/>
```

[a]
```
/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/lcg/root/6.26.11-149c0390ac099fe3c8283ca35bade3b4/bin/genreflex src/DataFormats/PortableTestObjects/src/alpaka/classes_rocm.h --debug  -s src/DataFormats/PortableTestObjects/src/alpaka/classes_rocm_def.xml -o tmp/el8_amd64_gcc11/src/DataFormats/PortableTestObjects/src/alpaka/DataFormatsPortableTestObjectsROCmAsync/a/DataFormatsPortableTestObjectsROCmAsync_xr.cc --fail_on_warnings --rootmap=tmp/el8_amd64_gcc11/src/DataFormats/PortableTestObjects/src/alpaka/DataFormatsPortableTestObjectsROCmAsync/a/DataFormatsPortableTestObjectsROCmAsync_xr.rootmap --rootmap-lib=libDataFormatsPortableTestObjectsROCmAsync.so -m DataFormatsCommon_xr_rdict.pcm -m DataFormatsProvenance_xr_rdict.pcm -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DCMS_DICT_IMPL -D_REENTRANT -DGNUSOURCE -D__STRICT_ANSI__ -DALPAKA_ACC_GPU_HIP_ENABLED -DALPAKA_HOST_ONLY -DGNU_GCC -D_GNU_SOURCE -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DEIGEN_DONT_PARALLELIZE -DTBB_USE_GLIBCXX_VERSION=110201 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH="CMSSW_13_1_X_2023-03-16-2300" -DPROJECT_NAME="CMSSW" -DPROJECT_VERSION="CMSSW_13_1_X_2023-03-16-2300" -I/build/muz/del/x/CMSSW_13_1_X_2023-03-16-2300/src -I/build/muz/del/x/CMSSW_13_1_X_2023-03-16-2300/poison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc11/cms/cmssw/CMSSW_13_1_X_2023-03-16-2300/src -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/alpaka/develop-20230215-009369d4b30704f235d4afa876f7f9b0/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/pcre/8.43-5dcc901acc02f624b22dd9840b2357e8/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/boost/1.80.0-165bd6bfbfb4303b695a8c5946a5b92e/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/bz2lib/1.0.6-2c1f18484cb66c30aba7929f2be5e7d4/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/libuuid/2.34-0451b31e1b9a58c6aeefab41c18eea34/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm-rocrand/5.4.3-5cda51a0a8ef64a6f63f35d330d3072d/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/lcg/root/6.26.11-149c0390ac099fe3c8283ca35bade3b4/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/tbb/v2021.8.0-75c29f2d261a174db9e74a1b202d0dd2/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/xz/5.2.5-83d0a00b575efd1701e07bedf7977343/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/zlib/1.2.11-3dfb2715f3608466b74431b80eb9d788/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/eigen/82dd3710dac619448f50331c1d6a35da673f764a-9ac4aed18ac60d0189693c592862694d/include/eigen3 -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/fmt/8.0.1-43b841663c2a0d6622910a1ad66d228d/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/md5/1.0.0-e68283f2de2e2e709a0db99db3b53205/include -isystem/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/rocm/5.4.3-e875674b47f88526f6f9c4594f5cafdd/include -I/cvmfs/cms-ci.cern.ch/week0/PR_60c12b0d/el8_amd64_gcc11/external/tinyxml2/6.2.0-c2bad61e58f94d6db8f640afbd739be2/include -DCMSSW_REFLEX_DICT
```

**Comment by vgvassilev** [comment]
Hi @smuzaffar, indeed it seems that genreflex does not support this feature. I think we can open a feature request for it. However, in a longer run, we should move to `rootcling` as it is the main tool (genreflex partially forwards to it). I believe that will not be very expensive in terms of efforts as we have done it already for the CXXMODULE IB.

**Comment by smuzaffar** [comment]
thanks @vgvassilev , sure we can move to `rootcling`.  I can update our rules to use `rootcling`. Most of the command-line option are already there but I noticed that `genreflex` also passes `-v4 -inlineInputHeader` to `rootcling` . Should I also include these in direct call to `rootcling`? How can I know with -`vN` I should use OR `-v4` is for debugging?

**Comment by vgvassilev** [comment]
> thanks @vgvassilev , sure we can move to `rootcling`. I can update our rules to use `rootcling`. Most of the command-line option are already there but I noticed that `genreflex` also passes `-v4 -inlineInputHeader` to `rootcling` .

I think we will need at least `-inlineInputHeader` to preserve the original behavior. Maybe we could at a later stage remove it if it is not needed anymore...

> Should I also include these in direct call to `rootcling`? How can I know with -`vN` I should use OR `-v4` is for debugging?

The default is `-v2` IIRC, `-v4` is for debugging. The good thing about `roocling` is that it has better structured documentation, ie. `rootcling --help` is useful sometimes :)

**Comment by fwyzard** [comment]
Does `rootcling` support `-isystem` ?
`rootcling --help` does not mention it.

**Comment by vgvassilev** [comment]
> Does `rootcling` support `-isystem` ? `rootcling --help` does not mention it.

I think this is the relevant option `  --compilerI=<string>     - Specify a compiler default include path, to suppress unneeded "-isystem" arguments.`

**Comment by smuzaffar** [comment]
@fwyzard , we will still need to add `rocm-rocrand` include directory in to `ROOT_INCLUDE_PATH`  for runtime. IIRC, root still needs to parse the header at runtime and for that it uses `ROOT_INCLUDE_PATH` to find the headers

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
> ---> test test-das-selected-lumis had ERRORS
> ---> test test_edmPickEvents had ERRORS
> ---> test DRNTest had ERRORS

@smuzaffar should I investigate these or are they know failures ?

**Comment by smuzaffar** [comment]
@fwyzard , no need to investigate these. `test-das-selected-lumis and  test_edmPickEvents ` fail due to missing el9 cms env for this week and ` DRNTest` has the same old issue that it randomly fails on different nodes because it failed to start a container within a container

**Comment by smuzaffar** [comment]
by the way, this will fail for rhel7 ... right? Should we enable rocm only for rhel8 and 9?

**Comment by fwyzard** [comment]
do we still have el7 builds for 13.1.x ?
if we do, let's test it, it should work.

**Comment by fwyzard** [comment]
please test for slc7_amd64_gcc11

**Comment by smuzaffar** [comment]
yes we do have slc7 as most of the lxplus resources are based on centos7. By the end of the year lxplus will become alma9 so may be at that time we can drop slc7 or we can disucss in core sw meeting it we want to keep slc7 alive.
By the way I do not see https://repo.radeon.com/rocm/rhel7 that is why I think rocm will fail to download rhel7 rpms

**Comment by fwyzard** [comment]
Ah, I see... the el7 repository is https://repo.radeon.com/rocm/yum/ , I'll adjust the spec files.

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
please test for slc7_amd64_gcc11

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
> \+ rpm2cpio /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/SOURCES/external/rocm/5.4.3-6547a6b81235d5c321f6f1d1aaeadaac/comgr-2.4.0.50403-121.el7.x86_64.rpm
> \+ cpio -idmv
>rpm2cpio: error while loading shared libraries: libdb-6.0.so: cannot open shared object file: No such file or directory
>cpio: premature end of archive
>error: Bad exit status from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/rpm-tmp.O95VPr (%build)

@smuzaffar , ideas ?

**Comment by smuzaffar** [comment]
@fwyzard , we have https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_13_1_X/master/rpm-preamble.file#L164 to protect cms packages picking up libs from bootstrap bundle. For `slc7` this causes `rpm2cpio` tool to fail which needed bootstrap bundle libs. `rpm2cpio` works in `el8/el9` as rpm package is built with `rpath` but `slc7` has an old rpm build and `rpm2cpio` there still requires bootstrap bundle `LD_LIBRARY_PATH`.  I would suggest to add
```
%undefine drop_bootstrap_lib
%define drop_bootstrap_lib true
```
in the `%if "%{rhel}" == "7"` block

**Comment by fwyzard** [comment]
Thanks, will try.

**Comment by smuzaffar** [comment]
please test for slc7_amd64_gcc11

**Comment by fwyzard** [comment]
please test for el9_amd64_gcc11

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
The same failures are present in last night's IB.

**Comment by fwyzard** [comment]
In the 13.0.x version (https://github.com/cms-sw/cmsdist/pull/8371) all unit tests pass.

**Comment by fwyzard** [comment]
@smuzaffar @perrotta @rappoccio can we merge this for CMSSW_13_1_0_pre2 ?
  - all unit test failures are already present in the nightly IB
  - the 13.0.x backport passes all tests

**Comment by smuzaffar** [comment]
+externals

**Comment by perrotta** [comment]
+orp

**Comment by perrotta** [comment]
merge

**Pull request merged**